common/elasticsearch/interfaces.go

```diff
@@ -0,0 +1,225 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package elasticsearch
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/uber/cadence/common/metrics"
+
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/log"
+	p "github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/service/config"
+)
+
+// NewGenericElasticSearchClient create a ES client
+func NewGenericElasticSearchClient(
+	connectConfig *config.ElasticSearchConfig,
+	visibilityConfig *config.VisibilityConfig,
+	logger log.Logger,
+) (GenericElasticSearch, error) {
+	// TODO hardcoded to V6 for now
+	return newV6Client(connectConfig, visibilityConfig, logger)
+}
+
+type (
+	// GenericElasticSearch is a generic interface for all versions of ElasticSearch clients
+	GenericElasticSearch interface {
+		Search(ctx context.Context, request *SearchRequest) (*p.InternalListWorkflowExecutionsResponse, error)
+		SearchByQuery(ctx context.Context, request *SearchByQueryRequest) (*p.InternalListWorkflowExecutionsResponse, error)
+		ScanByQuery(ctx context.Context, request *ScanByQueryRequest) (*p.InternalListWorkflowExecutionsResponse, error)
+		CountByQuery(ctx context.Context, index, query string) (int64, error)
+		PutMapping(ctx context.Context, index, root, key, valueType string) error
+		CreateIndex(ctx context.Context, index string) error
+		GetClosedWorkflowExecution(ctx context.Context, index string, request *p.InternalGetClosedWorkflowExecutionRequest) (*p.InternalGetClosedWorkflowExecutionResponse, error)
+
+		RunBulkProcessor(ctx context.Context, p *GenericBulkProcessorParameters) (GenericBulkProcessor, error)
+	}
+
+	// SearchRequest is request for Search
+	SearchRequest struct {
+		Index       string
+		ListRequest *p.InternalListWorkflowExecutionsRequest
+		IsOpen      bool
+		Filter      IsRecordValidFilter
+		MatchQuery  *GenericMatch
+	}
+
+	// GenericMatch is a match struct
+	GenericMatch struct {
+		Name string
+		Text interface{}
+	}
+
+	// SearchByQueryRequest is request for SearchByQuery
+	SearchByQueryRequest struct {
+		Index         string
+		Query         string
+		NextPageToken []byte
+		PageSize      int
+		Filter        IsRecordValidFilter
+	}
+
+	// ScanByQueryRequest is request for SearchByQuery
+	ScanByQueryRequest struct {
+		Index         string
+		Query         string
+		NextPageToken []byte
+		PageSize      int
+	}
+
+	// GenericBulkProcessor is a bulk processor
+	GenericBulkProcessor interface {
```

Comment:
Maybe just `BulkProcessor`? I think if a definition lives in this file, it implies it's "Generic" to all elastic version, so probably we can drop all "Generic" in the interface or struct definition?

Reply:
I am fine with naming, because it's better to let more people be happy with names to maintain the code. 
I am actually coming from another side that I feel like making this explicitly "Generic" is better. But again this is very minor things.

Reply:
Let's keep the name to be consistent with `GenericClient`.